### DISCUSSION TITLE
Questions about launching a similar project

### DISCUSSION THREAD

**Discussion by tylermumford** [casey/just#1593] [General]
Hi, @casey and the Just community. I've created an open source CLI tool for developers. (Not linking to it unless asked.) I was inspired by Just and other tools like it, and I have some questions about launching and running a project in this category.

1. Did you do anything to "spread the word" about Just? Or did it just "take off" once people started discovering it on their own?
2. Which package managers are the most important to support? I'm planning to support Homebrew and Scoop, but I'm sure people wanted Just to be available in lots of ways.
3. Are any of Just's features "traps" or rabbit holes that take a lot more effort than you expected?
4. What makes it sustainable? What do you do that makes this project fun or rewarding or meaningful?

Thanks for creating an awesome tool I love to use. And thanks for writing [this blog post]; I really liked the section on testing.

[this blog post]: https://rodarmor.com/blog/tour-de-just/

**Comment by casey** [comment]
Hi Tyler!

Great questions. And feel free to link, I don't mind at all.

> 1. Did you do anything to "spread the word" about Just? Or did it just "take off" once people started discovering it on their own?

I've made a bunch of posts to different subreddits and to Hacker News over the years, which I think has been responsible for getting the initial critical mass of users. These days, Just seems to grow on its own through organic growth, word of mouth, and users making blog posts.

> 2. Which package managers are the most important to support? I'm planning to support Homebrew and Scoop, but I'm sure people wanted Just to be available in lots of ways.

I've actually found that it's best to let users package your program. Once you have users, they'll package it themselves for the package managers they use, so I think your best bet is focusing on finding users and making your program useful. Plus, I've found that packaging Just myself takes a ton of time and effort, whereas for the users who package Just for a package manager, they often maintain a bunch of packages for that package manager already, so it's not a huge effort to package one more.

> 3. Are any of Just's features "traps" or rabbit holes that take a lot more effort than you expected?

Hmmm, good question. Just has depended on Cygwin on Windows in the past, and still does to a limited extent. This has been problematic, not because Cygwin is bad, but just because any external dependency makes a program way harder for users to run. I'd say try very hard to not have any external dependencies, e.g., dependencies that aren't a part of the program binary.

> 4. What makes it sustainable? What do you do that makes this project fun or rewarding or meaningful?

I've actually gotten burnt out on working on Just a bunch of times. However, when that happens I'll just stop working on it for a while, or drop a feature I was working on, and eventually I'll be less burnt out and I'll want to work on it again. I've never accepted donations or payment for working on Just, and although there are some features that would be nice to have, Just works great as is. All that means that I just feel free to ignore it when I don't want to work on it.

> Thanks for creating an awesome tool I love to use. And thanks for writing [this blog post](https://rodarmor.com/blog/tour-de-just/); I really liked the section on testing.

You bet! And extensive tests are *huge* for maintainability. I'll often start working on Just after a long break, start adding a feature, which would break something, but which is caught by a test before I try to merge it. This saves me from having to think super super hard about what the implications of making a change are a lot of the time, because there'll usually be a test that catches problems before they happen.

Best,
Casey

**Reply by tylermumford** [reply]
Thank you so much! These answers are thoughtful and super helpful.

I'm creating a program called [LocalStatus] for monitoring finicky/untracked aspects of local dev environments. It's super helpful for me at work, and it's been fun to work on.

[LocalStatus]: https://github.com/tylermumford/localstatus

> ... any external dependency makes a program way harder for users to run. I'd say try very hard to not have any external dependencies, e.g., dependencies that aren't a part of the program binary.

This is absolutely true. Unfortunately, I may be stepping directly into the problem already. 😅 My tool runs the git and npm programs to do some of its checks. I'm hoping it won't be a problem because they're super common dev tools and not required for the core functionality of my tool, but... I'm going to have to think about it. Maybe there's another way.

Thanks again for such a helpful reply! Live long and prosper. 😊
